File: Top-Interview-Questions-Leetcode/Easy-Collections/dynamic-programming/best-time-to-buy-and-sell-stock.py
from typing import List

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        mn_price, mx_profit = prices[0], 0
        
        for i in prices:
            mx_profit = max(mx_profit, i - mn_price)
            mn_price = min(mn_price, i)
        
        return mx_profit

File: Top-Interview-Questions-Leetcode/Easy-Collections/dynamic-programming/best-time-to-buy-and-sell-stock.cpp
#include <iostream>
#include <vector>
#include <climits>

class Solution {
public:
    int maxProfit(std::vector<int>& prices) {
        
        int mx_profit = 0, mn_price = INT_MAX;
        
        for (int i : prices) {
            mx_profit = std::max(mx_profit, i - mn_price);
            mn_price = std::min(mn_price, i);
        }
        
        return mx_profit;
    }
};